### PR TITLE
Implementation of crosstalk for ALLEGRO ECAL Barrel

### PR DESCRIPTION
**Pull request by zwu0922**
**The following PR of k4FWCore has to be processed before this update:**
https://github.com/key4hep/k4FWCore/pull/193

New features:
1. A class for reading the crosstalk map. https://github.com/zwu0922/k4RecCalorimeter/blob/main/RecCalorimeter/src/components/ReadCaloCrosstalkMap.h
2. The crosstalk is added to the existing "CreateCaloCells" class. The computation of crosstalk can be switched off to speed up the simulation. https://github.com/zwu0922/k4RecCalorimeter/blob/main/RecCalorimeter/src/components/CreateCaloCells.h
3. Steering file. https://github.com/zwu0922/k4RecCalorimeter/blob/main/RecCalorimeter/tests/options/runEcalBarrel_ReconstructionTopoClusters_crosstalk.py
The output file "output_test_EcalBarrel_crosstalk.root" contains the topo-clusters with crosstalk effect for 5 events.

### CODE REVIEW

File: RecCalorimeter/src/components/CreateCaloCells.cpp [hunk 1]
```diff
@@ -84,13 +90,41 @@ StatusCode CreateCaloCells::execute() {
     m_cellsMap.clear();
   }
 
+  m_HitCellsMap.clear();
```
Comment (kjvbrt) [inline RecCalorimeter/src/components/CreateCaloCells.cpp:93]: Can you make `m_HitCellsMap` just a local variable? This way you will not have to clear it in every event.

**Comment by kjvbrt** [comment]
Hi @zwu0922, can you update your branch to the latest upstream?

**Comment by zwu0922** [comment]
> Hi @zwu0922, can you update your branch to the latest upstream?

Done. No conflict needs to be resolved

File: RecCalorimeter/src/components/CreateCaloCells.cpp (hunk 1 above)
Comment (zwu0922) [reply RecCalorimeter/src/components/CreateCaloCells.cpp]: I changed the name of the variable to "HitCellsMap" and made it local

**Comment by zwu0922** [comment]
> > Hi @zwu0922, can you update your branch to the latest upstream?
> 
> Done. No conflict needs to be resolved

Looks like the test does not include some recent development in k4FWCore?
https://github.com/key4hep/k4FWCore/commit/5f9a69f2ad3b44fc3cc34979ea9f364999c37e56

**Comment by kjvbrt** [comment]
Hi @zwu0922 that is not the case, I'm able to compile your PR locally.

The issue seems to be coming from the `VolumeManager`

```
CreateECalBarre...WARNING Size of sampling fraction values is smaller than the number of existing layers. Taking the sa
mpling fraction for last layer. Layer ID: 11                                                                           VolumeManager    ERROR lookupContext: Failed to search Volume context 0000004D2FF05804 [Unknown identifier]            
CreateECalBarre...  FATAL  Standard std::exception is caught                                                           
CreateECalBarre...  ERROR VolumeManager: lookupContext: Failed to search Volume context 0000004D2FF05804 [Unknown ident
ifier]                                                                                                                 
EventLoopMgr        FATAL .executeEvent(): Standard std::exception thrown by CreateECalBarrelCells                     
EventLoopMgr        ERROR VolumeManager: lookupContext: Failed to search Volume context 0000004D2FF05804 [Unknown ident
ifier]
```

**Comment by giovannimarchiori** [comment]
are you using ALLEGRO v03 (11 layers in ECAL barrel) or v01,v02 (12 layers)?

**Comment by zwu0922** [comment]
> are you using ALLEGRO v03 (11 layers in ECAL barrel) or v01,v02 (12 layers)?

I used ALLEGRO v02 (12 layers) to test my code during the development.

**Comment by kjvbrt** [comment]
The test which fails is:
```
RecFCCeeCalorimeter/tests/options/run_thetamodulemerged.py
```
With
```
FCCee/ALLEGRO/compact/ALLEGRO_o1_v03/ALLEGRO_o1_v03.xml
```

File: RecCalorimeter/tests/options/runEcalBarrel_ReconstructionTopoClusters_crosstalk.py
Comment (kjvbrt) [inline RecCalorimeter/tests/options/runEcalBarrel_ReconstructionTopoClusters_crosstalk.py:1]: Can you register this test in `RecFCCeeCalorimeter/CMakeLists.txt`?

**Comment by giovannimarchiori** [comment]
> > are you using ALLEGRO v03 (11 layers in ECAL barrel) or v01,v02 (12 layers)?
> 
> I used ALLEGRO v02 (12 layers) to test my code during the development.
We have now moved to ALLEGRO v03 (11 layers, aligned cell corners) as baseline

File: RecCalorimeter/tests/options/runEcalBarrel_ReconstructionTopoClusters_crosstalk.py
Comment (zwu0922) [reply RecCalorimeter/tests/options/runEcalBarrel_ReconstructionTopoClusters_crosstalk.py:1]: I added the test at the end of CMakeLists.txt

**Comment by zwu0922** [comment]
> > > are you using ALLEGRO v03 (11 layers in ECAL barrel) or v01,v02 (12 layers)?
> > 
> > 
> > I used ALLEGRO v02 (12 layers) to test my code during the development.
> > We have now moved to ALLEGRO v03 (11 layers, aligned cell corners) as baseline

Thanks, I see. The crosstalk computation does not depend on the number of layers. But of course, I will update the input to the test to match the ALLEGRO v03 design afterwards.

**Comment by kjvbrt** [comment]
Hi @zwu0922, can you adjust the input to the `FCCeeLAr_simRecoAllegroV3` test in this PR?

Also, your test `FCCeeLAr_runxtalk` depends on a variable defined only in `LAr_Scripts`, can you remove this dependency?

```
Test #8: FCCeeLAr_runxtalk ................***Failed    1.42 sec
/cvmfs/sw-nightlies.hsf.org/key4hep/releases/2024-06-20/x86_64-almalinux9-gcc11.4.1-opt/k4geo/77e501e47007088f1770eb7009cf1b3684b009f1_develop-4yk2ue/share/k4geo
Traceback (most recent call last):
  File "/home/jsmiesko/zwu0922/k4FWCore/install/bin/k4run", line 241, in <module>
    main()
  File "/home/jsmiesko/zwu0922/k4FWCore/install/bin/k4run", line 185, in main
    load_file(file)
  File "/home/jsmiesko/zwu0922/k4FWCore/install/python/k4FWCore/utils.py", line 51, in load_file
    exec(code, globals())
  File "RecCalorimeter/tests/options/runEcalBarrel_ReconstructionTopoClusters_crosstalk.py", line 260, in <module>
    fileName=os.environ['FCCBASEDIR'] + neighboursMap,
  File "/cvmfs/sw-nightlies.hsf.org/key4hep/releases/2024-05-09/x86_64-almalinux9-gcc11.4.1-opt/python/3.10.13-vrpxgy/lib/python3.10/os.py", line 680, in __getitem__
    raise KeyError(key) from None
KeyError: 'FCCBASEDIR'
```

**Comment by BrieucF** [comment]
> Hi @zwu0922, can you adjust the input to the `FCCeeLAr_simRecoAllegroV3` test in this PR?
> 
> Also, your test `FCCeeLAr_runxtalk` depends on a variable defined only in `LAr_Scripts`, can you remove this dependency?
> 
> ```
> Test #8: FCCeeLAr_runxtalk ................***Failed    1.42 sec
> /cvmfs/sw-nightlies.hsf.org/key4hep/releases/2024-06-20/x86_64-almalinux9-gcc11.4.1-opt/k4geo/77e501e47007088f1770eb7009cf1b3684b009f1_develop-4yk2ue/share/k4geo
> Traceback (most recent call last):
>   File "/home/jsmiesko/zwu0922/k4FWCore/install/bin/k4run", line 241, in <module>
>     main()
>   File "/home/jsmiesko/zwu0922/k4FWCore/install/bin/k4run", line 185, in main
>     load_file(file)
>   File "/home/jsmiesko/zwu0922/k4FWCore/install/python/k4FWCore/utils.py", line 51, in load_file
>     exec(code, globals())
>   File "RecCalorimeter/tests/options/runEcalBarrel_ReconstructionTopoClusters_crosstalk.py", line 260, in <module>
>     fileName=os.environ['FCCBASEDIR'] + neighboursMap,
>   File "/cvmfs/sw-nightlies.hsf.org/key4hep/releases/2024-05-09/x86_64-almalinux9-gcc11.4.1-opt/python/3.10.13-vrpxgy/lib/python3.10/os.py", line 680, in __getitem__
>     raise KeyError(key) from None
> KeyError: 'FCCBASEDIR'
> ```

For ALLEGRO v03, the files needed for reconstruction can be found here `/eos/project/f/fccsw-web/www/filesForSimDigiReco/ALLEGRO/ALLEGRO_o1_v03/` (see https://github.com/HEP-FCC/FCC-config/tree/main/FCCee/FullSim/ALLEGRO/ALLEGRO_o1_v03#running-the-digitization-and-reconstruction). 

Let me know where are the files you need, I will upload them there.

And as you pointed out in a previous discussion, for the tests, you should be able to directly access these files with something like  `fileName = "https://fccsw.web.cern.ch/fccsw/filesForSimDigiReco/ALLEGRO/ALLEGRO_o1_v03/neighbours_map_ecalB_thetamodulemerged.root"`

**Comment by BrieucF** [comment]
Hi @zwu0922 shall we finalize this?

**Comment by zwu0922** [comment]
> Hi @zwu0922 shall we finalize this?

Sorry for the huge delay. I had a problem running the new ddsim for ALLEGRO v03 (Switching off the HCAL simulation always leads to a crash). That is why I didn't manage to adapt my development to ALLEGRO v03. I will contact Tong and others to sort it out as soon as possible and finalise the implementation of cross-talk emulation.

**Comment by BrieucF** [comment]
Ok thanks, let me know when I can copy your files

**Comment by zwu0922** [comment]
> > > are you using ALLEGRO v03 (11 layers in ECAL barrel) or v01,v02 (12 layers)?
> > 
> > 
> > I used ALLEGRO v02 (12 layers) to test my code during the development.
> > We have now moved to ALLEGRO v03 (11 layers, aligned cell corners) as baseline

Thanks for the notice. I have adapted the test macro to ALLEGRO v03

**Comment by zwu0922** [comment]
> Hi @zwu0922, can you adjust the input to the `FCCeeLAr_simRecoAllegroV3` test in this PR?
> 
> Also, your test `FCCeeLAr_runxtalk` depends on a variable defined only in `LAr_Scripts`, can you remove this dependency?
> 
> ```
> Test #8: FCCeeLAr_runxtalk ................***Failed    1.42 sec
> /cvmfs/sw-nightlies.hsf.org/key4hep/releases/2024-06-20/x86_64-almalinux9-gcc11.4.1-opt/k4geo/77e501e47007088f1770eb7009cf1b3684b009f1_develop-4yk2ue/share/k4geo
> Traceback (most recent call last):
>   File "/home/jsmiesko/zwu0922/k4FWCore/install/bin/k4run", line 241, in <module>
>     main()
>   File "/home/jsmiesko/zwu0922/k4FWCore/install/bin/k4run", line 185, in main
>     load_file(file)
>   File "/home/jsmiesko/zwu0922/k4FWCore/install/python/k4FWCore/utils.py", line 51, in load_file
>     exec(code, globals())
>   File "RecCalorimeter/tests/options/runEcalBarrel_ReconstructionTopoClusters_crosstalk.py", line 260, in <module>
>     fileName=os.environ['FCCBASEDIR'] + neighboursMap,
>   File "/cvmfs/sw-nightlies.hsf.org/key4hep/releases/2024-05-09/x86_64-almalinux9-gcc11.4.1-opt/python/3.10.13-vrpxgy/lib/python3.10/os.py", line 680, in __getitem__
>     raise KeyError(key) from None
> KeyError: 'FCCBASEDIR'
> ```

I replaced the test macro for the cross-talk calculation by a modification of the digitisation script for ALLEGRO v03. There is no dependence on `LAr_Scripts`.

**Comment by zwu0922** [comment]
> > Hi @zwu0922, can you adjust the input to the `FCCeeLAr_simRecoAllegroV3` test in this PR?
> > Also, your test `FCCeeLAr_runxtalk` depends on a variable defined only in `LAr_Scripts`, can you remove this dependency?
> > ```
> > Test #8: FCCeeLAr_runxtalk ................***Failed    1.42 sec
> > /cvmfs/sw-nightlies.hsf.org/key4hep/releases/2024-06-20/x86_64-almalinux9-gcc11.4.1-opt/k4geo/77e501e47007088f1770eb7009cf1b3684b009f1_develop-4yk2ue/share/k4geo
> > Traceback (most recent call last):
> >   File "/home/jsmiesko/zwu0922/k4FWCore/install/bin/k4run", line 241, in <module>
> >     main()
> >   File "/home/jsmiesko/zwu0922/k4FWCore/install/bin/k4run", line 185, in main
> >     load_file(file)
> >   File "/home/jsmiesko/zwu0922/k4FWCore/install/python/k4FWCore/utils.py", line 51, in load_file
> >     exec(code, globals())
> >   File "RecCalorimeter/tests/options/runEcalBarrel_ReconstructionTopoClusters_crosstalk.py", line 260, in <module>
> >     fileName=os.environ['FCCBASEDIR'] + neighboursMap,
> >   File "/cvmfs/sw-nightlies.hsf.org/key4hep/releases/2024-05-09/x86_64-almalinux9-gcc11.4.1-opt/python/3.10.13-vrpxgy/lib/python3.10/os.py", line 680, in __getitem__
> >     raise KeyError(key) from None
> > KeyError: 'FCCBASEDIR'
> > ```
> 
> For ALLEGRO v03, the files needed for reconstruction can be found here `/eos/project/f/fccsw-web/www/filesForSimDigiReco/ALLEGRO/ALLEGRO_o1_v03/` (see https://github.com/HEP-FCC/FCC-config/tree/main/FCCee/FullSim/ALLEGRO/ALLEGRO_o1_v03#running-the-digitization-and-reconstruction).
> 
> Let me know where are the files you need, I will upload them there.
> 
> And as you pointed out in a previous discussion, for the tests, you should be able to directly access these files with something like `fileName = "https://fccsw.web.cern.ch/fccsw/filesForSimDigiReco/ALLEGRO/ALLEGRO_o1_v03/neighbours_map_ecalB_thetamodulemerged.root"`

Thanks. Now I point all input root files in the test macro to the central directory for ALLEGRO v03.

**Pull request merged**